src/encoder.rs

```diff
@@ -24,140 +21,150 @@ use byteorder::{BigEndian, ByteOrder};
+///     }
+/// );
+/// assert!(encoder::encode_into(&packet, &mut bytes).is_ok())
+/// ```
+pub fn encode_into<O: Output>(packet: &OscPacket, out: &mut O) -> Result<usize, O::Err> {
     match *packet {
-        OscPacket::Message(ref msg) => encode_message(msg),
-        OscPacket::Bundle(ref bundle) => encode_bundle(bundle),
+        OscPacket::Message(ref msg) => encode_message(msg, out),
+        OscPacket::Bundle(ref bundle) => encode_bundle(bundle, out),
     }
 }
 
-fn encode_message(msg: &OscMessage) -> Result<Vec<u8>> {
-    let mut msg_bytes: Vec<u8> = Vec::new();
-
-    msg_bytes.extend(encode_string(msg.addr.clone()));
-    let mut type_tags: Vec<char> = vec![','];
-    let mut arg_bytes: Vec<u8> = Vec::new();
+fn encode_message<O: Output>(msg: &OscMessage, out: &mut O) -> Result<usize, O::Err> {
+    let mut written = encode_string_into(&msg.addr, out)?;
 
+    written += out.write(b",")?;
     for arg in &msg.args {
-        let (bytes, tags): (Option<Vec<u8>>, String) = encode_arg(arg)?;
-
-        type_tags.extend(tags.chars());
-        if let Some(data) = bytes {
-            arg_bytes.extend(data);
-        }
+        written += encode_arg_type(arg, out)?;
     }
 
-    msg_bytes.extend(encode_string(type_tags.into_iter().collect::<String>()));
-    if !arg_bytes.is_empty() {
-        msg_bytes.extend(arg_bytes);
-    }
-    Ok(msg_bytes)
-}
-
-fn encode_bundle(bundle: &OscBundle) -> Result<Vec<u8>> {
-    let mut bundle_bytes: Vec<u8> = Vec::new();
-    bundle_bytes.extend(encode_string("#bundle".to_string()).into_iter());
+    let padding = pad(written as u64 + 1) as usize - written;
+    written += out.write(&[0u8; 4][..padding])?;
 
-    match encode_arg(&OscType::Time(bundle.timetag))? {
-        (Some(x), _) => {
-            bundle_bytes.extend(x.into_iter());
-        }
-        (None, _) => {
-            return Err(OscError::BadBundle("Missing time tag!".to_string()));
-        }
+    for arg in &msg.args {
+        written += encode_arg_data(arg, out)?;
     }
 
-    if bundle.content.is_empty() {
-        return Ok(bundle_bytes);
-    }
+    Ok(written)
+}
+
+fn encode_bundle<O: Output>(bundle: &OscBundle, out: &mut O) -> Result<usize, O::Err> {
+    let mut written = encode_string_into("#bundle", out)?;
+    written += encode_time_tag_into(&bundle.timetag, out)?;
 
     for packet in &bundle.content {
         match *packet {
             OscPacket::Message(ref m) => {
-                let msg = encode_message(m)?;
-                let mut msg_size = vec![0u8; 4];
-                BigEndian::write_u32(&mut msg_size, msg.len() as u32);
-                bundle_bytes.extend(msg_size.into_iter().chain(msg.into_iter()));
+                let msg_len = encode_message(m, &mut NullOutput).unwrap();
+                out.reserve(4 + msg_len)?;
+                written += out.write(&(msg_len as u32).to_be_bytes())?;
+                written += encode_message(m, out)?;
```

Comment:
It's not ideal that we have to encode the message twice here, first to get the length just by discarding the encoding output, and second for the actual encoding.  How much effort do you think would it be to encode just once?

Note that I don't see this as a hard requirement, because this is something that could be done as a follow-up.

Reply:
I agree, that additionally requiring the `Seek` trait would limit the general usability.

> Generics are monomorphized, so encoding into a NullOutput should effectively be compiled into just calculating and returning the size.

Good point, that should actually be the case.

> If you've got any suggestions on a better way to do this, feel free.

Not yet, but thank you for pointing out that a future change might be a breaking one.  So, I'd like to not merge it now, and instead reflect on this in a couple of days again.   Your performance argument is valid, it might not be worth to optimize this.

Reply:
Thinking on this, I see two possible approaches for improving this that could theoretically improve performance, while still allowing the output to be written in a single pass (i.e. without seeking):

1. Encode messages/bundles into a temporary Vec, then write out the length and the message. This would prevent the necessary double-traversal of the packet hierarchy, but would reintroduce a temporary buffer. Using them in this limited capacity could prove beneficial.
2. Memoize the size calculation (if still doing a size pre-pass) or packet encoding (if using buffered encoding a la point 1). This could potentially help with deep hierarchies in both cases, as each OscPacket would only ever be encoded once. This would also require a heap allocation, but for a different purpose (probably using a `HashMap<*const OscPacket, usize>` or something along those lines). I think the biggest improvement from this would be with a deep hierarchy (e.g. `bundle(bundle(bundle(...)))`) due to the O(n^2) visits of the current approach, so I may add an additional benchmark for that as well.

I'll try making these changes and benchmarking them to see whether either of these changes could provide additional performance.

I'll also add here that it occurred to me that unless explicitly replaced (e.g. with `mimalloc` or `jemalloc`), Rust uses the system allocator. Thus, the operating system can have a large impact on the performance of allocation. The numbers currently in the table in the PR description were run on a Windows installation. To compare, I gave the benchmarks a preliminary run on my Linux dual-boot as well (same hardware). The overall performance improvement still looks pretty significant across the board, but it does appear that allocation isn't quite as costly on Linux. The benchmarks that reuse a Vec were nearly the same as on Windows , while all other benchmarks (which allocated new Vecs) were significantly faster than in my original runs on Windows.

I'll be updating the table in the PR description with columns for runtimes on Linux on the same hardware when I get a chance, but I may not be able to get to either task (points above or Linux benchmarks) until tomorrow.

Reply:
Added a benchmark for nested bundles. The performance is definitely pathological in the case of a deeply-nested hierarchy.

Encoding bundle contents into a temporary Vec results in comparable performance overall, possibly even slightly beneficial depending on the packet in question. In the benchmarks, it was slower in the case with a wide bundle, faster in the case of the huge bundle, and does completely solve the pathological case. The changes were pushed to the `encoder-improvements+temp` branch on my fork. This seems like a good solution.

Memoizing the sizes also solves the pathological case, but performance overall is worse than just using a temporary Vec. This would require an additional dependency anyway for the `no_std` case, as there is no `HashMap` in `alloc` (as noted before), so `hashbrown` would need to be pulled in. The changes were pushed to the `encoder-improvements+memoize` branch on my fork. This solution doesn't seem practical.

Thinking on it more, doing both memoization and encoding into a temporary Vec isn't beneficial; the O(n^2) nature arises from having a separate size pass, so if a temporary encoding is done, memoization is no longer beneficial.

A comparison of the possible solutions:

| Benchmark | `encoder-improvements` | `encoder-improvements+temp` | `encoder-improvements+memoize` |
|---|---|---|---|
| bench_encode_bundles | 6,331 ± 15 | 9,211 ± 45 | 22,880 ± 702 |
| bench_encode_huge_bundle | 223,253 ± 977 | 156,850 ± 2,248 | 254,628 ± 5,354 |
| bench_encode_messages | 11,307 ± 52 | 11,790 ± 34 | 28,292 ± 126 |
| bench_encode_nested_bundles | 3,787,008 ± 12,171 | 1,624 ± 8 | 1,550 ± 21 |

I think encoding bundle contents into a temporary Vec is the best solution. Temporary buffers are good in moderation!

Reply:
Any additional thoughts on this?

Reply:
Sorry for the delay, I'll try to review this again this weekend.

Reply:
Thank you for the clear and detailed explanation of the potential performance improvements!  There's not much to add from my side, except that I agree with your statement that encoding into a temporary `Vec` looks like the best choice.  It's also simpler than the memoization approach.

I could make a benchmark run on macOS as well if you like, but I guess that the performance will be similar to the Linux results.

The only thing that is left to do, is to squash the commits.  What do you think of squashing everything that's related to benchmarking into one commit and everything performance related into another one?  It must not be exactly two commits at the end, but I think it would be nice to have the whole performance improvement patches in one commit.

Reply:
I agree that the MacOS performance is probably similar to the Linux performance, but it certainly can't hurt to check. You don't necessarily have to format it into a table like I have been (it's a pain to do manually and I still haven't created any kind of shortcut) but a quick check that there aren't any regressions would be good.

As far as commits go, squash away. I definitely tended towards a "commit early and often" approach here, so the number of commits is a bit high. I'm OK with a benchmarks commit and a code changes commit, though if you do the benchmarks first the `encode_into` benchmarks will break. Changes then benchmarks should be good, though.

Thank you for reviewing and thank you for the crate!

Reply:
> You don't necessarily have to format it into a table like I have been (it's a pain to do manually and I still haven't created any kind of shortcut) but a quick check that there aren't any regressions would be good.

There definitely is no regression, but I also did not expect one.

Benchmark for first commit of this branch  (36b080f): [bench-36b080f.log](https://github.com/klingtnet/rosc/files/11079170/bench-36b080f.log)
Benchmark for the latest commit of this branch (b7b18bc): [bench-b7b18bc.log](https://github.com/klingtnet/rosc/files/11079171/bench-b7b18bc.log)

If you need any other benchmarks, please let me know.  They were done with Rust `1.70.0-nightly` on a M1 MacBook Air.

>  I definitely tended towards a "commit early and often" approach here, so the number of commits is a bit high.

That's totally fine and also the way I usually work.  But, eventually we only care about the end result, without all the intermediate steps.  You're also free to keep any intermediate step/commit where you think it's worth keeping it (for reference, future investigation, debugging, etc).  Any cleanups, refactorings etc. should be squashed or fixed up, though.

> Thank you for reviewing and thank you for the crate!

You're welcome.  I'd like to spend more time on the maintenance of this, and some other crates, but I cannot do that as much as I'd like.  So I'm very happy to receive PRs like yours.  This keeps this project alive.

Reply:
Sorry for the delay! This slipped my mind on Monday, and now that I'm starting to look at it again, I'm thinking there's a better solution.

I'm starting to think it would be worth bringing back the write-back mechanism on `Output`. I had done that in my original pass of the `Output` trait, as a means of avoiding a temporary allocation with the single-scan encoding, then made it multi-scan to avoid the need for it. Now that we've effectively gone back to single-scan, though, I'm thinking that adding back a write-back mechanism would be a better alternative than doing internal temporary allocations as we have it now. It increases the requirements of the `Output` trait (effectively requiring it to be seekable), but if the alternative is internally allocating a buffer, I feel that should probably be left in the user's hands. The user can very easily encode into a buffer themselves using `encode` and then write to whatever non-seekable output is needed, but for suitable outputs (e.g. appending to an existing buffer) that should make `encode_into` more efficient without introducing the pathological case that nested bundles had with the multi-scan encoding.

This should also be somewhat more efficient than creating internal buffers, as a single buffer can cover the entire packet, rather than one per message/bundle. Going to play around with this a bit more.

Sorry for changing my mind and putting this back in motion after calling it done!

Reply:
Implemented this as described above, it seems to have resulted in *very slight* improvements to most benchmarks on Windows. Benchmarks which encode a lot of bundles (`encode_bundles` and `encode_nested_bundles`) improved substantially.

There's a bit of a regression in the huge bundle benchmarks allocating a new vector compared to the temporary buffer implementation, but a decent gain in the benchmark reusing a vector.

Seems like an overall win to me. Will test on Linux as well.

To allow for the use of `Seek + Write` types as `Output`s, I've added a `WriteOutput<W>` newtype, which can be used as a wrapper to allow any `Seek + Write` type to be used as an `Output`, while still providing a `Vec<u8>` implementation.

| Benchmark                                | Temporary Buffer (Windows) | Write-back (Windows) | Temporary Buffer (Linux) |
|------------------------------------------|----------------------------|----------------------|--------------------------|
| bench_encode_args_array                  | 6,683 ± 138                | 5,929 ± 51           | 5,220 ± 53               |
| bench_encode_args_blob                   | 10,401 ± 676               | 8,940 ± 69           | 10,098 ± 43              |
| bench_encode_args_bool                   | 5,426 ± 36                 | 4,932 ± 107          | 4,992 ± 28               |
| bench_encode_args_double                 | 6,078 ± 77                 | 5,506 ± 39           | 5,104 ± 31               |
| bench_encode_args_float                  | 5,875 ± 68                 | 5,413 ± 22           | 4,760 ± 58               |
| bench_encode_args_int                    | 5,944 ± 39                 | 5,751 ± 43           | 4,847 ± 33               |
| bench_encode_args_long                   | 5,983 ± 176                | 5,425 ± 46           | 5,200 ± 52               |
| bench_encode_args_nil                    | 5,237 ± 35                 | 4,983 ± 47           | 4,749 ± 28               |
| bench_encode_args_string                 | 10,148 ± 95                | 9,083 ± 50           | 9,650 ± 41               |
| bench_encode_bundles                     | 10,044 ± 82                | 5,844 ± 48           | 8,759 ± 104              |
| bench_encode_bundles_into_new_vec        | 11,236 ± 155               | 6,825 ± 70           | 10,076 ± 49              |
| bench_encode_bundles_into_reused_vec     | 9,983 ± 46                 | 5,282 ± 52           | 9,712 ± 144              |
| bench_encode_huge_bundle                 | 511,412 ± 212,302          | 562,780 ± 11,300     | 144,198 ± 1,989          |
| bench_encode_huge_bundle_into_new_vec    | 435,239 ± 17,202           | 565,861 ± 13,714     | 151,398 ± 673            |
| bench_encode_huge_bundle_into_reused_vec | 162,223 ± 2,096            | 137,702 ± 713        | 153,435 ± 3,161          |
| bench_encode_messages                    | 12,695 ± 287               | 10,903 ± 241         | 11,017 ± 49              |
| bench_encode_messages_into_new_vec       | 15,146 ± 310               | 11,988 ± 55          | 12,697 ± 68              |
| bench_encode_messages_into_reused_vec    | 13,728 ± 175               | 10,386 ± 54          | 12,343 ± 86              |
| bench_encode_nested_bundles              | 4,561 ± 82                 | 640 ± 25             | 1,530 ± 23               |